### PR TITLE
Remove N-number of samples from `getSample` and change to `Vector{T}`

### PR DESCRIPTION
**Pull request by Affie**
- [x] Remove N from `getSample`
- [x] update to `Vector{T}` where T<:Tuple for now, but I don't think we should limit it in the future \
- [x] Make ccw.measurement type stable
- [x] fix almost everything above broke 
- [x] Gradients

TODO
- [ ] Circular (Should be changed to CircularGroup)
- [ ] Fix DERelitive use of sampleFactor to getSample

### CODE REVIEW

**Comment by Affie** [comment]
Hi @dehann, I hope this is on the right track for getSample update? There is still a lot to do, but it will only take time.

File: src/Factors/GenericFunctions.jl [hunk 1]
```diff
@@ -55,12 +55,16 @@ end
 
 DFG.getManifold(f::ManifoldFactor) = f.M
 
-function getSample(cf::CalcFactor{<:ManifoldFactor}, N::Int=1)
+function getSample(cf::CalcFactor{<:ManifoldFactor{M,Z}}) where {M,Z}
     #TODO @assert dim == cf.factor.Z's dimension
     #TODO investigate use of SVector if small dims
-    ret = [rand(cf.factor.Z) for _ in 1:N]
+    if M isa ManifoldKernelDensity
```
Comment (dehann) [inline src/Factors/GenericFunctions.jl:61]: suggest doing this with dispatch while figuring out `sample`, `rand`, `getSample`, `sampleTangent`

File: src/services/EvalFactor.jl [hunk 2]
```diff
@@ -424,7 +424,7 @@ end
 function evalPotentialSpecific( Xi::AbstractVector{<:DFGVariable},
                                 ccwl::CommonConvWrapper{Mixture{N_,F,S,T}},
                                 solvefor::Symbol,
-                                measurement::Tuple=(Vector{Vector{Float64}}(),);
+                                measurement::Vector{<:Tuple}=Tuple[];#FIXME
```
Comment (dehann) [inline src/services/EvalFactor.jl:427]:
```suggestion
                                measurement::AbstractVector{<:Tuple}=Tuple[];#FIXME
```

File: src/services/EvalFactor.jl [hunk 3]
```diff
@@ -438,7 +438,7 @@ end
 function evalPotentialSpecific( Xi::AbstractVector{<:DFGVariable},
                                 ccwl::CommonConvWrapper{F},
                                 solvefor::Symbol,
-                                measurement::Tuple=(Vector{Vector{Float64}}(),);
+                                measurement::Vector{<:Tuple}=Tuple[];#FIXME
```
Comment (dehann) [inline src/services/EvalFactor.jl:441]:
```suggestion
                                measurement::AbstractVector{<:Tuple}=Tuple[];#FIXME
```

File: src/services/EvalFactor.jl [hunk 4]
```diff
@@ -458,10 +458,10 @@ Single entry point for evaluating factors from factor graph, using multiple disp
 function evalFactor(dfg::AbstractDFG,
                     fct::DFGFactor,
                     solvefor::Symbol,
-                    measurement::Tuple=(Vector{Vector{Float64}}(),);
+                    measurement::Vector{<:Tuple}=Tuple[];#FIXME
```
Comment (dehann) [inline src/services/EvalFactor.jl:461]:
```suggestion
                    measurement::AbstractVector{<:Tuple}=Tuple[];#FIXME
```

File: src/services/EvalFactor.jl [hunk 5]
```diff
@@ -511,7 +511,7 @@ Related
 function _evalFactorTemporary!( fct::AbstractFactor,
                                 varTypes::Tuple,
                                 sfidx::Int,  # solve for index, assuming variable order for fct
-                                meas_single::Tuple,
+                                measurement::Vector{<:Tuple},
```
Comment (dehann) [inline src/services/EvalFactor.jl:514]:
```suggestion
                                measurement::AbstractVector{<:Tuple},
```

File: src/services/FactorGradients.jl [hunk 6]
```diff
@@ -7,7 +7,7 @@ export checkGradientsToleranceMask, calcPerturbationFromVariable
 # T_pt_args[:] = [(T1::Type{<:InferenceVariable}, point1); ...]
 # FORCED TO START AT EITHER :x1
 function _prepFactorGradientLambdas(fct::Union{<:AbstractRelativeMinimize,<:AbstractRelativeRoots, <:AbstractManifoldMinimize}, 
-                                    measurement::Tuple,
+                                    measurement::Vector{<:Tuple},
```
Comment (dehann) [inline src/services/FactorGradients.jl:10]:
```suggestion
                                    measurement::AbstractVector{<:Tuple},
```

File: src/services/FactorGradients.jl [hunk 7]
```diff
@@ -75,7 +75,7 @@ end
 
 function FactorGradientsCached!(fct::Union{<:AbstractRelativeMinimize, <:AbstractRelativeRoots, <:AbstractManifoldMinimize},
                                 varTypes::Tuple,
-                                meas_single::Tuple, 
+                                meas_single::Vector{<:Tuple},
```
Comment (dehann) [inline src/services/FactorGradients.jl:78]:
```suggestion
                                meas_single::AbstractVector{<:Tuple}, 
```

File: src/DeconvUtils.jl [hunk 8]
```diff
@@ -39,7 +39,7 @@ Related
 function approxDeconv(fcto::DFGFactor,
                       ccw::CommonConvWrapper = _getCCW(fcto);
                       N::Int=100,
-                      measurement::Tuple=sampleFactor(ccw, N),
+                      measurement::Vector{<:Tuple}=sampleFactor(ccw, N),
```
Comment (dehann) [inline src/DeconvUtils.jl:42]:
```suggestion
                      measurement::AbstractVector{<:Tuple}=sampleFactor(ccw, N),
```

File: src/entities/FactorGradients.jl [hunk 9]
```diff
@@ -58,7 +58,7 @@ Related
 
 [`calcFactorResidualTemporary`](@ref), [`_buildGraphByFactorAndTypes`](@ref)
 """
-mutable struct FactorGradientsCached!{F <: AbstractRelative,S,M<:Tuple,P,G,L}
+mutable struct FactorGradientsCached!{F <: AbstractRelative, S, M <: Vector{<:Tuple}, P, G, L}
```
Comment (dehann) [inline src/entities/FactorGradients.jl:61]: can perhaps move to `M <: AbstractVector{<:Tuple}` later, just leaving a breadcrum.

File: src/services/ApproxConv.jl [hunk 10]
```diff
@@ -6,9 +6,9 @@ export calcFactorResidual
 function approxConvBelief(dfg::AbstractDFG,
                           fc::DFGFactor,
                           target::Symbol,
-                          measurement::Tuple=(Vector{Vector{Float64}}(),);
+                          measurement::Vector{<:Tuple}=Tuple[];#FIXME
```
Comment (dehann) [inline src/services/ApproxConv.jl:9]:
```suggestion
                          measurement::AbstractVector{<:Tuple}=Tuple[];#FIXME
```

File: src/services/ApproxConv.jl [hunk 11]
```diff
@@ -62,9 +62,9 @@ Related
 function approxConvBelief(dfg::AbstractDFG, 
                           from::Symbol, 
                           target::Symbol,
-                          measurement::Tuple=(Vector{Vector{Float64}}(),);
+                          measurement::Vector{<:Tuple}=Tuple[];#FIXME
```
Comment (dehann) [inline src/services/ApproxConv.jl:65]:
```suggestion
                          measurement::AbstractVector{<:Tuple}=Tuple[];#FIXME
```

File: src/services/ApproxConv.jl [hunk 12]
```diff
@@ -152,8 +152,8 @@ Notes
 function calcProposalBelief(dfg::AbstractDFG,
                             fct::DFGFactor,
                             target::Symbol,
-                            measurement::Tuple=(Vector{Vector{Float64}}(),);
-                            N::Int=length(measurement[1]),
+                            measurement::Vector{<:Tuple}=Tuple[];#FIXME
```
Comment (dehann) [inline src/services/ApproxConv.jl:155]:
```suggestion
                            measurement::AbstractVector{<:Tuple}=Tuple[];#FIXME
```

File: src/services/ApproxConv.jl [hunk 13]
```diff
@@ -168,8 +168,8 @@ end
 function calcProposalBelief(dfg::AbstractDFG,
                             fct::DFGFactor{<:CommonConvWrapper{<:PartialPriorPassThrough}},
                             target::Symbol,
-                            measurement::Tuple=(Vector{Vector{Float64}}(),);
-                            N::Int=length(measurement[1]),
+                            measurement::Vector{<:Tuple}=Tuple[];#FIXME
```
Comment (dehann) [inline src/services/ApproxConv.jl:171]:
```suggestion
                            measurement::AbstractVector{<:Tuple}=Tuple[];#FIXME
```

File: src/services/ApproxConv.jl [hunk 14]
```diff
@@ -207,7 +207,7 @@ function proposalbeliefs!(dfg::AbstractDFG,
                           destlbl::Symbol,
                           factors::AbstractVector{<:DFGFactor},
                           dens::AbstractVector{<:ManifoldKernelDensity},
-                          measurement::Tuple=(Vector{Vector{Float64}}(),);
+                          measurement::Vector{<:Tuple}=Tuple[];#FIXME
```
Comment (dehann) [inline src/services/ApproxConv.jl:210]:
```suggestion
                          measurement::AbstractVector{<:Tuple}=Tuple[];#FIXME
```

File: src/services/CalcFactor.jl [hunk 15]
```diff
@@ -118,7 +118,7 @@ Related
 """
 function calcFactorResidualTemporary( fct::AbstractRelative, 
                                       varTypes::Tuple,
-                                      measurement::Tuple,
+                                      measurement::Vector{<:Tuple},
```
Comment (dehann) [inline src/services/CalcFactor.jl:121]:
```suggestion
                                      measurement::AbstractVector{<:Tuple},
```

File: src/services/EvalFactor.jl (hunk 2 above)
Comment (Affie) [reply src/services/EvalFactor.jl:427]: Hi, I don't have any local changes, was going to spend some time on it now, so let me know when you are done and we can merge alll the review comments before I look

File: src/services/EvalFactor.jl [hunk 16]
```diff
@@ -257,22 +257,22 @@ function evalPotentialSpecific( Xi::AbstractVector{<:DFGVariable},
                                 ccwl::CommonConvWrapper{T},
                                 solvefor::Symbol,
                                 T_::Type{<:AbstractRelative},
-                                measurement::Tuple=(Vector{Float64}(),);
+                                measurement::Vector{<:Tuple}=Tuple[];#FIXME
```
Comment (dehann) [inline src/services/EvalFactor.jl:260]:
```suggestion
                                measurement::AbstractVector{<:Tuple}=Tuple[];#FIXME
```

File: src/services/EvalFactor.jl [hunk 17]
```diff
@@ -322,10 +322,10 @@ function evalPotentialSpecific( Xi::AbstractVector{<:DFGVariable},
                                 ccwl::CommonConvWrapper{T},
                                 solvefor::Symbol,
                                 T_::Type{<:AbstractPrior},
-                                measurement::Tuple=(Vector{Vector{Float64}}(),);
+                                measurement::Vector{<:Tuple}=Tuple[];#FIXME
```
Comment (dehann) [inline src/services/EvalFactor.jl:325]:
```suggestion
                                measurement::AbstractVector{<:Tuple}=Tuple[];#FIXME
```

**Review by dehann — APPROVED** [review]
see comments, looks good to me thanks!

**Comment by Affie** [comment]
Cool, I'll just work on fixing the tests and will update. will merge your review suggestions now.

File: src/Factors/GenericFunctions.jl (hunk 1 above)
Comment (Affie) [reply src/Factors/GenericFunctions.jl:61]: It's a bit of an odd one out, so I quickly did the if to get it to run. I still want to try and avoid 2 `getSample` functions on this one. still TODO

Comment (dehann) [reply src/Factors/GenericFunctions.jl:61]: Sounds good

**Pull request merged**